File: 2019/src/test/java/aoc/Day18Test.java
package aoc;

import org.junit.jupiter.api.*;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;

import static org.junit.jupiter.api.Assertions.assertEquals;

class Day18Test {
    private static String example;
    private static String example2;
    private static String example3;
    private static String example4;
    private static String example5;
    private static String example6;
    private static String example7;
    private static String example8;
    private static String example9;
    private static String input;
    private Day18 day;

    @BeforeAll
    public static void init() throws IOException {
        example = Files.readString(Paths.get("src/main/resources/day18_example.txt"), StandardCharsets.UTF_8);
        example2 = Files.readString(Paths.get("src/main/resources/day18_example2.txt"), StandardCharsets.UTF_8);
        example3 = Files.readString(Paths.get("src/main/resources/day18_example3.txt"), StandardCharsets.UTF_8);
        example4 = Files.readString(Paths.get("src/main/resources/day18_example4.txt"), StandardCharsets.UTF_8);
        example5 = Files.readString(Paths.get("src/main/resources/day18_example5.txt"), StandardCharsets.UTF_8);
        example6 = Files.readString(Paths.get("src/main/resources/day18_example6.txt"), StandardCharsets.UTF_8);
        example7 = Files.readString(Paths.get("src/main/resources/day18_example7.txt"), StandardCharsets.UTF_8);
        example8 = Files.readString(Paths.get("src/main/resources/day18_example8.txt"), StandardCharsets.UTF_8);
        example9 = Files.readString(Paths.get("src/main/resources/day18_example9.txt"), StandardCharsets.UTF_8);
        input = Files.readString(Paths.get("src/main/resources/day18.txt"), StandardCharsets.UTF_8);
    }

    @BeforeEach
    public void setup() {
        day = new Day18();
    }

    @Nested
    @DisplayName("Part 1")
    class Part1Tests {
        @Test
        void testExample() {
            long result = day.part1(example);
            assertEquals(8L, result);
        }

        @Test
        void testExample2() {
            long result = day.part1(example2);
            assertEquals(86L, result);
        }

        @Test
        void testExample3() {
            long result = day.part1(example3);
            assertEquals(132L, result);
        }

        @Test
        void testExample4() {
            long result = day.part1(example4);
            assertEquals(136L, result);
        }

        @Test
        void testExample5() {
            long result = day.part1(example5);
            assertEquals(81L, result);
        }

        @Test
        void testSolution() {
            long result = day.part1(input);
            assertEquals(7430L, result);
        }
    }

    @Nested
    @DisplayName("Part 2")
    class Part2Tests {
        @Test
        void testExample6() {
            long result = day.part2(example6);
            assertEquals(8L, result);
        }

        @Test
        void testExample7() {
            long result = day.part2(example7);
            assertEquals(24L, result);
        }

        @Test
        void testExample8() {
            long result = day.part2(example8);
            assertEquals(32L, result);
        }

        @Test
        void testExample9() {
            long result = day.part2(example9);
            assertEquals(72L, result);
        }

        @Test
        void testSolution() {
            long result = day.part2(input);
            assertEquals(1864L, result);
        }
    }
}

File: 2020/aoc2020/src/test/kotlin/Day18Test.kt
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals

@DisplayName("Day 18")
class Day18Test {
    private val example1 = readText("day18_example1.txt")
    private val example2 = readText("day18_example2.txt")
    private val example3 = readText("day18_example3.txt")
    private val example4 = readText("day18_example4.txt")
    private val example5 = readText("day18_example5.txt")
    private val example6 = readText("day18_example6.txt")
    private val input = readText("day18.txt")
    private val day = Day18()

    @Nested
    @DisplayName("Part 1")
    inner class Part1 {
        @Test
        fun `Matches example 1`() {
            val answer = day.part1(example1)
            assertEquals(71, answer)
        }

        @Test
        fun `Matches example 2`() {
            val answer = day.part1(example2)
            assertEquals(51, answer)
        }

        @Test
        fun `Matches example 3`() {
            val answer = day.part1(example3)
            assertEquals(26, answer)
        }

        @Test
        fun `Matches example 4`() {
            val answer = day.part1(example4)
            assertEquals(437, answer)
        }

        @Test
        fun `Matches example 5`() {
            val answer = day.part1(example5)
            assertEquals(12240, answer)
        }

        @Test
        fun `Matches example 6`() {
            val answer = day.part1(example6)
            assertEquals(13632, answer)
        }

        @Test
        fun `Matches solution`() {
            val answer = day.part1(input)
            assertEquals(6_640_667_297_513, answer)
        }
    }

    @Nested
    @DisplayName("Part 2")
    inner class Part2 {
        @Test
        fun `Matches example 1`() {
            val answer = day.part2(example1)
            assertEquals(231, answer)
        }

        @Test
        fun `Matches example 2`() {
            val answer = day.part2(example2)
            assertEquals(51, answer)
        }

        @Test
        fun `Matches example 3`() {
            val answer = day.part2(example3)
            assertEquals(46, answer)
        }

        @Test
        fun `Matches example 4`() {
            val answer = day.part2(example4)
            assertEquals(1445, answer)
        }

        @Test
        fun `Matches example 5`() {
            val answer = day.part2(example5)
            assertEquals(669060, answer)
        }

        @Test
        fun `Matches example 6`() {
            val answer = day.part2(example6)
            assertEquals(23340, answer)
        }

        @Test
        fun `Matches solution`() {
            val answer = day.part2(input)
            assertEquals(451_589_894_841_552, answer)
        }
    }
}
